PANDAS

To represent tabular data, pandas uses a custom data structure called a dataframe. A dataframe is a highly efficient, 2-dimensional data structure that provides a suite of methods and attributes to quickly explore, analyze, and visualize data. The dataframe is similar to the NumPy 2D array but adds support for many features that help you work with tabular data.

One of the biggest advantages that pandas has over NumPy is the ability to store mixed data types in rows and columns. Many tabular datasets contain a range of data types and pandas dataframes handle mixed data types effortlessly while NumPy doesn't. Pandas dataframes can also handle missing values gracefully using a custom object, NaN, to represent those values. A common complaint with NumPy is its lack of an object to represent missing values and people end up having to find and replace these values manually. In addition, pandas dataframes contain axis labels for both rows and columns and enable you to refer to elements in the dataframe more intuitively. Since many tabular datasets contain column titles, this means that dataframes preserve the metadata from the file around the data.

In this tutorial, we'll use Pandas to analyze data on video game reviews.
In order to be able to work with the data in Python, we'll need to read the csv file into a Pandas DataFrame. A DataFrame is a way to represent and work with tabular data. Tabular data has rows and columns, just like our csv file.
In order to read in the data, we'll need to use the pandas.read_csv function. This function will take in a csv file and return a DataFrame. The below code will:

In [44]:
import pandas as pd
reviews = pd.read_csv("C:\Users\hp\Documents\self.Python Tutorials\ign.csv")
# dataframe.columns will provide the column list of dataset as shown in code below:
print reviews.columns

Index([u'Unnamed: 0', u'score_phrase', u'title', u'url', u'platform', u'score',
       u'genre', u'editors_choice', u'release_year', u'release_month',
       u'release_day'],
      dtype='object')


Once we read in a DataFrame, Pandas gives us two methods that make it fast to print out the data. These functions are:
pandas.DataFrame.head -- prints the first N rows of a DataFrame. By default 5.
pandas.DataFrame.tail -- prints the last N rows of a DataFrame. By default 5.
We'll use the head method to see what's in reviews:

In [45]:
reviews.head()

Unnamed: 0 score_phrase                                              title  \
0           0      Amazing                            LittleBigPlanet PS Vita   
1           1      Amazing  LittleBigPlanet PS Vita -- Marvel Super Hero E...   
2           2        Great                               Splice: Tree of Life   
3           3        Great                                             NHL 13   
4           4        Great                                             NHL 13   

                                                 url          platform  score  \
0             /games/littlebigplanet-vita/vita-98907  PlayStation Vita    9.0   
1  /games/littlebigplanet-ps-vita-marvel-super-he...  PlayStation Vita    9.0   
2                          /games/splice/ipad-141070              iPad    8.5   
3                      /games/nhl-13/xbox-360-128182          Xbox 360    8.5   
4                           /games/nhl-13/ps3-128181     PlayStation 3    8.5   

        genre editors_choice  release_year  release_month  release_day  
0  Platformer              Y          2012              9           12  
1  Platformer              Y          2012              9           12  
2      Puzzle              N          2012              9           12  
3      Sports              N          2012              9           11  
4      Sports              N          2012              9           11

The columns contain information about that game:

score_phrase — how IGN described the game in one word. This is linked to the score it received.
title — the name of the game.
url — the URL where you can see the full review.
platform — the platform the game was reviewed on (PC, PS4, etc).
score — the score for the game, from 1.0 to 10.0.
genre — the genre of the game.
editors_choice — N if the game wasn't an editor's choice, Y if it was. This is tied to score.
release_year — the year the game was released.
release_month — the month the game was released.
release_day — the day the game was released.

In [46]:
reviews.shape

(18625, 11)

As you can see, everything has been read in properly -- we have 18625 rows and 11 columns.
One of the big advantages of Pandas vs just using NumPy is that Pandas allows you to have columns with different data types. reviews has columns that store float values, like score, string values, like score_phrase, and integers, like release_year.

In [47]:
 # Indexing DataFrames with Pandas using the pandas.DataFrame.iloc method
 # he below code will replicate reviews.head():
 reviews.iloc[:5, :]

Unnamed: 0 score_phrase                                              title  \
0           0      Amazing                            LittleBigPlanet PS Vita   
1           1      Amazing  LittleBigPlanet PS Vita -- Marvel Super Hero E...   
2           2        Great                               Splice: Tree of Life   
3           3        Great                                             NHL 13   
4           4        Great                                             NHL 13   

                                                 url          platform  score  \
0             /games/littlebigplanet-vita/vita-98907  PlayStation Vita    9.0   
1  /games/littlebigplanet-ps-vita-marvel-super-he...  PlayStation Vita    9.0   
2                          /games/splice/ipad-141070              iPad    8.5   
3                      /games/nhl-13/xbox-360-128182          Xbox 360    8.5   
4                           /games/nhl-13/ps3-128181     PlayStation 3    8.5   

        genre editors_choice  release_year  release_month  release_day  
0  Platformer              Y          2012              9           12  
1  Platformer              Y          2012              9           12  
2      Puzzle              N          2012              9           12  
3      Sports              N          2012              9           11  
4      Sports              N          2012              9           11

In [48]:
# Here are some indexing examples, along with the results:
# the entire DataFrame.
reviews.iloc[:,:]

Unnamed: 0 score_phrase  \
0               0      Amazing   
1               1      Amazing   
2               2        Great   
3               3        Great   
4               4        Great   
5               5         Good   
6               6        Awful   
7               7      Amazing   
8               8        Awful   
9               9         Good   
10             10         Good   
11             11         Good   
12             12         Good   
13             13      Amazing   
14             14      Amazing   
15             15         Okay   
16             16         Okay   
17             17        Great   
18             18     Mediocre   
19             19         Good   
20             20         Good   
21             21         Good   
22             22         Good   
23             23         Good   
24             24      Amazing   
25             25         Good   
26             26      Amazing   
27             27         Good   
28             28        Great   
29             29         Okay   
...           ...          ...   
18595       18595          Bad   
18596       18596         Okay   
18597       18597          Bad   
18598       18598         Okay   
18599       18599         Good   
18600       18600         Good   
18601       18601         Good   
18602       18602         Good   
18603       18603         Good   
18604       18604        Great   
18605       18605         Okay   
18606       18606         Okay   
18607       18607         Good   
18608       18608     Mediocre   
18609       18609        Great   
18610       18610         Okay   
18611       18611     Mediocre   
18612       18612         Good   
18613       18613        Great   
18614       18614      Amazing   
18615       18615      Amazing   
18616       18616         Good   
18617       18617        Great   
18618       18618      Amazing   
18619       18619         Good   
18620       18620         Good   
18621       18621      Amazing   
18622       18622     Mediocre   
18623       18623  Masterpiece   
18624       18624  Masterpiece   

                                                   title  \
0                                LittleBigPlanet PS Vita   
1      LittleBigPlanet PS Vita -- Marvel Super Hero E...   
2                                   Splice: Tree of Life   
3                                                 NHL 13   
4                                                 NHL 13   
5                              Total War Battles: Shogun   
6                                    Double Dragon: Neon   
7                                           Guild Wars 2   
8                                    Double Dragon: Neon   
9                              Total War Battles: Shogun   
10                               Tekken Tag Tournament 2   
11                               Tekken Tag Tournament 2   
12                                            Wild Blood   
13                                     Mark of the Ninja   
14                                     Mark of the Ninja   
15                       Home: A Unique Horror Adventure   
16                       Home: A Unique Horror Adventure   
17                                   Avengers Initiative   
18                                  Way of the Samurai 4   
19                           JoJo's Bizarre Adventure HD   
20                           JoJo's Bizarre Adventure HD   
21                              Mass Effect 3: Leviathan   
22                              Mass Effect 3: Leviathan   
23                              Mass Effect 3: Leviathan   
24                   Dark Souls (Prepare to Die Edition)   
25                                              Symphony   
26                                               Bastion   
27                     Tom Clancy's Ghost Recon Phantoms   
28                              Thirty Flights of Loving   
29                                             Legasista   
...                                   

In [49]:
# rows from position 5 onwards, and columns from position 5 onwards
reviews.iloc[5:, 5:]

score               genre editors_choice  release_year  release_month  \
5        7.0            Strategy              N          2012              9   
6        3.0            Fighting              N          2012              9   
7        9.0                 RPG              Y          2012              9   
8        3.0            Fighting              N          2012              9   
9        7.0            Strategy              N          2012              9   
10       7.5            Fighting              N          2012              9   
11       7.5            Fighting              N          2012              9   
12       7.0                 NaN              N          2012              9   
13       9.0   Action, Adventure              Y          2012              9   
14       9.0   Action, Adventure              Y          2012              9   
15       6.5           Adventure              N          2012              9   
16       6.5           Adventure              N          2012              9   
17       8.0              Action              N          2012              9   
18       5.5   Action, Adventure              N          2012              9   
19       7.0            Fighting              N          2012              9   
20       7.0            Fighting              N          2012              9   
21       7.5                 RPG              N          2012              8   
22       7.5                 RPG              N          2012              8   
23       7.5                 RPG              N          2012              8   
24       9.0         Action, RPG              Y          2012              8   
25       7.0             Shooter              N          2012              8   
26       9.0         Action, RPG              Y          2012              8   
27       7.5             Shooter              N          2012              8   
28       8.0           Adventure              N          2012              8   
29       6.5         Action, RPG              N          2012              8   
30       9.0           Adventure              Y          2012              8   
31       8.7                 RPG              Y          2012             10   
32       4.9          Platformer              N          2012             10   
33       9.6                 RPG              Y          2012             10   
34       7.3              Action              N          2012             10   
...      ...                 ...            ...           ...            ...   
18595    4.4              Action              N          2016              7   
18596    6.5   Action, Adventure              N          2016              7   
18597    4.9  Shooter, Adventure              N          2016              7   
18598    6.8              Action              N          2016              7   
18599    7.0              Battle              N          2016              7   
18600    7.4             Shooter              N          2016              8   
18601    7.4             Shooter              N          2016              8   
18602    7.4             Shooter              N          2016              8   
18603    7.8          Platformer              N          2016              8   
18604    8.6              Sports              N          2016              8   
18605    6.0           Adventure              N          2016              8   
18606    6.4            Strategy              N          2016              8   
18607    7.0              Battle              N          2016              7   
18608    5.4      Racing, Action              N          2016              7   
18609    8.0              Action              N          2016              7   
18610    6.0           Adventure              N          2016              7   
18611    5.8             Shooter              N          2016              7   
18612    7.8              Puzzle              N          2016              7   
18613  

In [50]:
# the first column, and all of the rows for the column.
reviews.iloc[:,0]

0            0
1            1
2            2
3            3
4            4
5            5
6            6
7            7
8            8
9            9
10          10
11          11
12          12
13          13
14          14
15          15
16          16
17          17
18          18
19          19
20          20
21          21
22          22
23          23
24          24
25          25
26          26
27          27
28          28
29          29
         ...  
18595    18595
18596    18596
18597    18597
18598    18598
18599    18599
18600    18600
18601    18601
18602    18602
18603    18603
18604    18604
18605    18605
18606    18606
18607    18607
18608    18608
18609    18609
18610    18610
18611    18611
18612    18612
18613    18613
18614    18614
18615    18615
18616    18616
18617    18617
18618    18618
18619    18619
18620    18620
18621    18621
18622    18622
18623    18623
18624    18624
Name: Unnamed: 0, Length: 18625, dtype: int64

In [51]:
# the 10th row, and all of the columns for that row.
reviews.iloc[9,:]

Unnamed: 0                                                9
score_phrase                                           Good
title                             Total War Battles: Shogun
url               /games/total-war-battles-shogun/pc-142564
platform                                                 PC
score                                                     7
genre                                              Strategy
editors_choice                                            N
release_year                                           2012
release_month                                             9
release_day                                              11
Name: 9, dtype: object

In [52]:
# Now that we know how to index by position, let's remove the first column, which doesn't have any useful information:
reviews = reviews.iloc[:, 1:]
reviews.head()

score_phrase                                              title  \
0      Amazing                            LittleBigPlanet PS Vita   
1      Amazing  LittleBigPlanet PS Vita -- Marvel Super Hero E...   
2        Great                               Splice: Tree of Life   
3        Great                                             NHL 13   
4        Great                                             NHL 13   

                                                 url          platform  score  \
0             /games/littlebigplanet-vita/vita-98907  PlayStation Vita    9.0   
1  /games/littlebigplanet-ps-vita-marvel-super-he...  PlayStation Vita    9.0   
2                          /games/splice/ipad-141070              iPad    8.5   
3                      /games/nhl-13/xbox-360-128182          Xbox 360    8.5   
4                           /games/nhl-13/ps3-128181     PlayStation 3    8.5   

        genre editors_choice  release_year  release_month  release_day  
0  Platformer              Y          2012              9           12  
1  Platformer              Y          2012              9           12  
2      Puzzle              N          2012              9           12  
3      Sports              N          2012              9           11  
4      Sports              N          2012              9           11

Pandas Series
The Series object is a core data structure that pandas uses to represent rows and columns. A Series is a labelled collection of values similar to the NumPy vector. The main advantage of Series objects is the ability to utilize non-integer labels. NumPy arrays can only utilize integer labels for indexing.

In [53]:
reviews["score"]

0         9.0
1         9.0
2         8.5
3         8.5
4         8.5
5         7.0
6         3.0
7         9.0
8         3.0
9         7.0
10        7.5
11        7.5
12        7.0
13        9.0
14        9.0
15        6.5
16        6.5
17        8.0
18        5.5
19        7.0
20        7.0
21        7.5
22        7.5
23        7.5
24        9.0
25        7.0
26        9.0
27        7.5
28        8.0
29        6.5
         ... 
18595     4.4
18596     6.5
18597     4.9
18598     6.8
18599     7.0
18600     7.4
18601     7.4
18602     7.4
18603     7.8
18604     8.6
18605     6.0
18606     6.4
18607     7.0
18608     5.4
18609     8.0
18610     6.0
18611     5.8
18612     7.8
18613     8.0
18614     9.2
18615     9.2
18616     7.5
18617     8.4
18618     9.1
18619     7.9
18620     7.6
18621     9.0
18622     5.8
18623    10.0
18624    10.0
Name: score, Length: 18625, dtype: float64

In [54]:
# We can also use lists of columns with this method:
reviews[["score", "release_year"]]

score  release_year
0        9.0          2012
1        9.0          2012
2        8.5          2012
3        8.5          2012
4        8.5          2012
5        7.0          2012
6        3.0          2012
7        9.0          2012
8        3.0          2012
9        7.0          2012
10       7.5          2012
11       7.5          2012
12       7.0          2012
13       9.0          2012
14       9.0          2012
15       6.5          2012
16       6.5          2012
17       8.0          2012
18       5.5          2012
19       7.0          2012
20       7.0          2012
21       7.5          2012
22       7.5          2012
23       7.5          2012
24       9.0          2012
25       7.0          2012
26       9.0          2012
27       7.5          2012
28       8.0          2012
29       6.5          2012
...      ...           ...
18595    4.4          2016
18596    6.5          2016
18597    4.9          2016
18598    6.8          2016
18599    7.0          2016
18600    7.4          2016
18601    7.4          2016
18602    7.4          2016
18603    7.8          2016
18604    8.6          2016
18605    6.0          2016
18606    6.4          2016
18607    7.0          2016
18608    5.4          2016
18609    8.0          2016
18610    6.0          2016
18611    5.8          2016
18612    7.8          2016
18613    8.0          2016
18614    9.2          2016
18615    9.2          2016
18616    7.5          2016
18617    8.4          2016
18618    9.1          2016
18619    7.9          2016
18620    7.6          2016
18621    9.0          2016
18622    5.8          2016
18623   10.0          2016
18624   10.0          2016

[18625 rows x 2 columns]

In [55]:
# We can verify that a single column is a Series:
type(reviews['score'])

pandas.core.series.Series

In [56]:
# We can create a Series manually to better understand how it works.
# To create a Series, we pass a list or NumPy array into the Series object when we instantiate it:
s1 = pd.Series([1,2])
s1

0    1
1    2
dtype: int64

In [57]:
s2 = pd.Series(['Pandas', 'are', 'better', 'than', 'Numpy', 99, 100])
s2

0    Pandas
1       are
2    better
3      than
4     Numpy
5        99
6       100
dtype: object

In [58]:
print s2[5]
type(s2[5])

99


int

In [59]:
print s2[2]
type(s2[2])

better


str

Creating A DataFrame in Pandas
We can create a DataFrame by passing multiple Series into the DataFrame class. Here, we pass in the two Series objects we just created, s1 as the first row, and s2 as the second row.

In [60]:
s2 = pd.Series(['Python', 'Programming'])
df = pd.DataFrame([s1,s2])
df.head()

0            1
0       1            2
1  Python  Programming

In [61]:
pd.DataFrame(
    [
        [1,2],
        ['Python', 'Programming']
    ]
)

0            1
0       1            2
1  Python  Programming

In [62]:
pd.DataFrame(
    [
        [1,2],
        ['Python', 'Programming']
    ],
    columns=["column1", "column2"]
)

column1      column2
0       1            2
1  Python  Programming

In [63]:
# Pandas DataFrame Methods
reviews["score"].mean()  # gets mean of column

6.950459060402685

In [64]:
# to get mean of each row
reviews.mean(axis=1)

0        510.500
1        510.500
2        510.375
3        510.125
4        510.125
5        509.750
6        508.750
7        510.250
8        508.750
9        509.750
10       509.875
11       509.875
12       509.500
13       509.250
14       509.250
15       508.375
16       508.375
17       508.500
18       507.375
19       507.750
20       507.750
21       514.625
22       514.625
23       514.625
24       515.000
25       514.250
26       514.750
27       514.125
28       514.250
29       513.625
          ...   
18595    510.850
18596    510.875
18597    510.225
18598    510.700
18599    510.750
18600    512.600
18601    512.600
18602    512.600
18603    512.450
18604    512.400
18605    511.500
18606    508.600
18607    510.750
18608    510.350
18609    510.750
18610    510.250
18611    508.700
18612    509.200
18613    508.000
18614    515.050
18615    515.050
18616    508.375
18617    508.600
18618    515.025
18619    514.725
18620    514.650
18621    515.000
18622    513.9

In [65]:
reviews['score'].count()

18625

In [66]:
print reviews.max()

score_phrase                       Unbearable
title                 xXx: State of the Union
url               /games/zusar-vasar/dc-13794
platform                                 iPod
score                                      10
genre                   Wrestling, Simulation
editors_choice                              Y
release_year                             2016
release_month                              12
release_day                                31
dtype: object


In [67]:
print reviews.min()

score_phrase                                     Amazing
title                                             #IDARB
url               /games/0-d-beat-drop/xbox-360-14342395
platform                                         Android
score                                                0.5
genre                                                inf
editors_choice                                         N
release_year                                        1970
release_month                                          1
release_day                                            1
dtype: object


In [68]:
print reviews.median()

score               7.3
release_year     2007.0
release_month       8.0
release_day        16.0
dtype: float64


In [69]:
print reviews.std()

score            1.711736
release_year     4.587529
release_month    3.476710
release_day      8.690128
dtype: float64


In [70]:
# DataFrame Math with Pandas
# All the common mathematical operators that work in Python, like +, -, *, /, and ^ will work, 
# and will apply to each element in a DataFrame or a Series.
reviews['score'] / 2

0        4.50
1        4.50
2        4.25
3        4.25
4        4.25
5        3.50
6        1.50
7        4.50
8        1.50
9        3.50
10       3.75
11       3.75
12       3.50
13       4.50
14       4.50
15       3.25
16       3.25
17       4.00
18       2.75
19       3.50
20       3.50
21       3.75
22       3.75
23       3.75
24       4.50
25       3.50
26       4.50
27       3.75
28       4.00
29       3.25
         ... 
18595    2.20
18596    3.25
18597    2.45
18598    3.40
18599    3.50
18600    3.70
18601    3.70
18602    3.70
18603    3.90
18604    4.30
18605    3.00
18606    3.20
18607    3.50
18608    2.70
18609    4.00
18610    3.00
18611    2.90
18612    3.90
18613    4.00
18614    4.60
18615    4.60
18616    3.75
18617    4.20
18618    4.55
18619    3.95
18620    3.80
18621    4.50
18622    2.90
18623    5.00
18624    5.00
Name: score, Length: 18625, dtype: float64

In [71]:
score_filter = reviews["score"] > 7
score_filter

0         True
1         True
2         True
3         True
4         True
5        False
6        False
7         True
8        False
9        False
10        True
11        True
12       False
13        True
14        True
15       False
16       False
17        True
18       False
19       False
20       False
21        True
22        True
23        True
24        True
25       False
26        True
27        True
28        True
29       False
         ...  
18595    False
18596    False
18597    False
18598    False
18599    False
18600     True
18601     True
18602     True
18603     True
18604     True
18605    False
18606    False
18607    False
18608    False
18609     True
18610    False
18611    False
18612     True
18613     True
18614     True
18615     True
18616     True
18617     True
18618     True
18619     True
18620     True
18621     True
18622    False
18623     True
18624     True
Name: score, Length: 18625, dtype: bool

In [72]:
filtered_reviews = reviews[score_filter]
filtered_reviews.head()

score_phrase                                              title  \
0      Amazing                            LittleBigPlanet PS Vita   
1      Amazing  LittleBigPlanet PS Vita -- Marvel Super Hero E...   
2        Great                               Splice: Tree of Life   
3        Great                                             NHL 13   
4        Great                                             NHL 13   

                                                 url          platform  score  \
0             /games/littlebigplanet-vita/vita-98907  PlayStation Vita    9.0   
1  /games/littlebigplanet-ps-vita-marvel-super-he...  PlayStation Vita    9.0   
2                          /games/splice/ipad-141070              iPad    8.5   
3                      /games/nhl-13/xbox-360-128182          Xbox 360    8.5   
4                           /games/nhl-13/ps3-128181     PlayStation 3    8.5   

        genre editors_choice  release_year  release_month  release_day  
0  Platformer              Y          2012              9           12  
1  Platformer              Y          2012              9           12  
2      Puzzle              N          2012              9           12  
3      Sports              N          2012              9           11  
4      Sports              N          2012              9           11

In [73]:
# When filtering with multiple conditions, it's important to put each condition in parentheses, 
# and separate them with a single ampersand (&).
xbox_one_filter = (reviews["score"] > 7) & (reviews["platform"] == "Xbox One")
filtered_reviews = reviews[xbox_one_filter]
filtered_reviews.head()

score_phrase                     title  \
17137      Amazing                 Gone Home   
17197      Amazing            Rayman Legends   
17295      Amazing  LEGO Marvel Super Heroes   
17313        Great             Dead Rising 3   
17317        Great           Killer Instinct   

                                                     url  platform  score  \
17137                 /games/gone-home/xbox-one-20014361  Xbox One    9.5   
17197            /games/rayman-legends/xbox-one-20008449  Xbox One    9.5   
17295  /games/lego-marvel-super-heroes/xbox-one-20000826  Xbox One    9.0   
17313               /games/dead-rising-3/xbox-one-124306  Xbox One    8.3   
17317      /games/killer-instinct-2013/xbox-one-20000538  Xbox One    8.4   

            genre editors_choice  release_year  release_month  release_day  
17137  Simulation              Y          2013              8           15  
17197  Platformer              Y          2013              8           26  
17295      Action              Y          2013             10           22  
17313      Action              N          2013             11           18  
17317    Fighting              N          2013             11           18